File: 剑指offer/链表/22.链表中倒数第k个节点/22.链表中倒数第k个节点.py
from util.leetcode_type import ListNode


class Solution:
    def getKthFromEnd(self, head: ListNode, k: int) -> ListNode:
        if not head:
            return head
        # fast 节点先前进 k
        # slow, fast同步前进，fast到队尾时 slow就是第k个节点
        fast, slow = head, head
        for i in range(k):
            fast = fast.next
        while fast:
            slow = slow.next
            fast = fast.next
        return slow

File: 剑指offer/链表/22.链表中倒数第k个节点/22.链表中倒数第k个节点.go
package 链表

import "algorithms-code-note/util/structures"

type ListNode = structures.ListNode

func getKthFromEnd(head *ListNode, k int) *ListNode {
	if head == nil {
		return head
	}
	fast := head
	slow := head
	for i := 0; i < k; i++ {
		fast = fast.Next
	}
	for fast != nil {
		fast = fast.Next
		slow = slow.Next
	}
	return slow
}
